### DISCUSSION TITLE
Scale overall robot in XML file without editing XML

### DISCUSSION THREAD

**Discussion by rohit-kumar-j** [google-deepmind/mujoco#330] [Asking for Help] [answered]
Can we scale the robot model in the XML file without editing the XML itself?

like:
```python
mujoco.MjModel.from_xml_path(
    file_name="path_to_xml", 
    globalScaling=10.0)
```

my robot needs to be twice the size according to the bounding box in fusion360. (The box is the bounding box)

**EDIT: The bounding box was not correctly specified**

<details>
    <Summary>Images</Summary>

---

This urdf export is from fusion with [fusion2urdf exporter](https://github.com/syuntoku14/fusion2urdf). This model is then converted to XML with compile.cc

`./compile path_to_urdf.urdf path_to_xml.xml`

<details>
    <Summary>In pybullet</Summary>

![image](https://user-images.githubusercontent.com/37873142/172632928-f51f992d-bb86-4add-ae1c-9a3d8f4d051e.png)

</details>

Inertias, masses, joints, etc have been exported properly, but not the meshes, they are half the size

<details>
    <Summary>In Mujoco</Summary>

![image](https://user-images.githubusercontent.com/37873142/172627918-9b318e5b-55d0-49a1-a6a1-660e9653ef46.png)

</details>

</details>

**Comment by nimrod-gileadi** [comment]
That's a good question. At the C API level, the route to creating a MuJoCo mjModel always goes through XML. Once an mjModel is created, scaling like that can't be done through the API. That means that scaling like you suggest does need to happen by creating a modified XML file, even if we hide it behind an API.

We've chosen to make the MuJoCo python bindings are a thin layer on top of the C APIs, so such functionality doesn't exist. Where it could live might be the [PyMJCF](https://github.com/deepmind/dm_control/blob/main/dm_control/mjcf/README.md) sub-library in dm_control.

Another thing we talked about was building a better [`<include>`](https://github.com/deepmind/dm_control/blob/main/dm_control/mjcf/README.md) tag which allows things like rescaling, coordinate transformations, and attaching to an existing kinematic tree.

For your case, I think the quickest route forward is to figure out why the export produces meshes of the wrong sizes, and make a fix either to fusion2urdf, or to MuJoCo's [URDF importer](https://github.com/deepmind/mujoco/blob/main/src/xml/xml_urdf.h), depending on the cause.

**Comment by rohit-kumar-j** [comment]
I have used [fusion2urdf](https://github.com/syuntoku14/fusion2urdf) and then [compile.cc](https://github.com/deepmind/mujoco/blob/main/sample/compile.cc) to convert it to an XML relatively easily.

However, upon conversion, the size of the robot in the urdf that was generated from fusion2urdf fit correctly when rendered in pybullet as was in the CAD software. When converting to XML using compile.cc,  the generated xml file was half the size.(1/8th in volume). Although the masses and the inertias were exported correctly.  

Note: The box is the approximate bounding box of the robot

<details>
   <Summary>In bullet</Summary>

![image](https://user-images.githubusercontent.com/37873142/173028866-d0f40d48-de57-40ac-a67a-f862b0a55928.png)

</details>

<details>
   <Summary>After compile.cc</Summary>

![image](https://user-images.githubusercontent.com/37873142/173029456-6083b618-7fc0-403b-b701-7e65cf9d05b4.png)

</details>

<details>
   <Summary>After compile.cc and scaled meshes to 2x</Summary>
Scaling meshes to 2x in `<assets></assets>`, the size fits, and as expected the joints are not correct

![image](https://user-images.githubusercontent.com/37873142/173029344-aad9c84e-3297-4101-a6f8-897a7b4108af.png)

**Comment by nimrod-gileadi** [comment]
OK. Given the same URDF works in Bullet, it's likely that there's a problem with MuJoCo's URDF importer.

Would you be able to share the generated URDF and maybe the assets?

**Comment by rohit-kumar-j** [comment]
I cannot give the source, however, I have replicated the same effect with a 4-bar mechanism. Here are the files:

[4_bar_scaling.zip](https://github.com/deepmind/mujoco/files/8878471/4_bar_scaling.zip)

There is a script to launch it in pybullet as well as one in mujoco.

In the produced XML (from `compile.cc`), I have only added the following lines for light, collisions and a bounding box of dims: 
1.05m x 0.15m x 1.05m

```xml
<option gravity="0 0 0" collision="predefined" />

<worldbody>    <!--within <worldbody>-->
    <light diffuse=".5 .5 .5" pos="0 -2 2" dir="0 1 -1" />
    <body pos="0 0 0">
        <geom type="box" size="1.05 0.15 1.05" rgba="1 1 1 .3" />
    </body>
</worldbody>
```
<details>
    <Summary>In pybullet</Summary>
Size: as defined in cad software

Scale: 1: 1 (if we assume 1=1m in pybullet) 

![image](https://user-images.githubusercontent.com/37873142/173057531-3f02808c-d96c-4b34-bbd0-726265720c12.png)

</details>

<details>
    <Summary>In mujoco</Summary>
Size: 1/2 in length, 1/8th in volume

Scale: 1: 0.5 (if we assume 1=1m in mujoco) 

![image](https://user-images.githubusercontent.com/37873142/173057589-39a3e079-3acd-4351-af90-48eb06917740.png)

</details>

**Comment by nimrod-gileadi** [comment, marked as answer]
`<geom type="box" size="1.05 0.15 1.05" rgba="1 1 1 .3" />`

defines a box that is 2.1m x 2.1m x 0.3m, since the size attribute on [`geom`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#body-geom) specifies half-size.

Does that explain the problem?

**Reply by yuvaltassa** [reply]
I suspect that indeed this was the issue. The URDF is imported correctly, you likely misinterpreted the size of your bounding box. Perhaps we should add this common gotcha (that in MuJoCo, geom sizes are like a radius, i.e. half-sizes), somewhere prominent in the modeling chapter.

**Reply by rohit-kumar-j** [reply]
Ah yes, this was the issue. My mistake!😊

**Reply by rohit-kumar-j** [reply]
shall I open an issue for this?
> Another thing we talked about was building a better [<include>](https://github.com/deepmind/dm_control/blob/main/dm_control/mjcf/README.md) tag which allows things like rescaling, coordinate transformations, and attaching to an existing kinematic tree.

**Reply by nimrod-gileadi** [reply]
It's very much on our mind but we need to articulate exactly what we want to do, so don't worry about it for now.

**Reply by rohit-kumar-j** [reply]
@nimrod-gileadi , I manually translated a `urdf` to an `xml` here: [pybullet_humanoid_for_mujoco](https://github.com/rohit-kumar-j/pybullet_humanoid_for_mujoco). (To use the [Human3.6M dataset](https://github.com/bulletphysics/bullet3/tree/master/examples/pybullet/gym/pybullet_envs/deep_mimic/mocap))

<details>
    <Summary>Human3.6M Walk</Summary>

https://user-images.githubusercontent.com/37873142/177263106-dabf8f17-f73c-4606-ae1c-7eaa9c25255b.mp4

</details>

When doing this, the things I had to scale down(or up) some of the the data sequentailly in the tags: 

```xml 
<geom: pos size />
<body: pos />
<inertial: pos /> <!-- Altough, this was eliminated as I didn't need inertial properties in the final xml -->
```

Although this is a simple example, perhaps could be loaded and manipulated like this:

```python
model = mujoco.MjModel.from_xml_path("robot.xml")
data = mujoco.MjData(model)

# via named access
data.body("top_level_body_name").scale= 1.25 # changes value of the pos, size while manintaing the origin in position
```

**Reply by nimrod-gileadi** [reply]
Nice work.

As I mentioned, scaling should be done on the model (rather than the data), and at the moment MuJoCo doesn't support such modifications once a model is "compiled" (i.e after from_xml_path). PyMJCF allows the modification of the XML structure before compilation, so probably scaling should go there.
